src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx

```diff
@@ -1158,15 +1162,25 @@ cpdef object compute_network_structured(
                 )
             
     # replace initial conditions with gage observations, wherever available
-    cdef int gages_size = usgs_positions_list.shape[0]
-    cdef int gage_i, usgs_position_i
+    cdef int gages_size = usgs_positions.shape[0]
     cdef int gage_maxtimestep = usgs_values.shape[1]
-
+    cdef int gage_i, usgs_position_i
+    cdef float a, da_decay_minutes, da_weight, da_shift, da_weighted_shift, replacement_value
+    cdef int [:] lastobs_timestep
+    cdef float [:] lastobs_values
     if gages_size:
+        lastobs_timestep = np.full(gages_size, -1, dtype='int32')
+        lastobs_values = np.zeros(gages_size, dtype='float32')
+        for gage_i in range(gages_size):
+            lastobs_values[gage_i] = lastobs_values_init[gage_i]
+
+    if gages_size and gage_maxtimestep > 0:
         for gage_i in range(gages_size):
-            usgs_position_i = usgs_positions_list[gage_i]
-            flowveldepth_nd[usgs_position_i, 0, 0] = usgs_values[gage_i, 0]
-            # TODO: handle the instance where there are no values, only gage positions
+            usgs_position_i = usgs_positions[gage_i]
+            # Handle the instance where there are no values, only gage positions
+            # TODO: Compare performance with math.isnan (imported for nogil...)
+            if not np.isnan(usgs_values[gage_i, 0]):
+                flowveldepth_nd[usgs_position_i, 0, 0] = usgs_values[gage_i, 0]
```

Comment:
This is the only line I was unsure about. This condition is meant to catch anytime there are no usgs values, but substitutes what in place for the first timestep? The empty array of usgs_values should not have anything. Is it just NaN? If it is not does the code below already cover this in line 1299?

Reply:
```
    if gages_size and gage_maxtimestep > 0:
```
if there is a valid last observation but no current usgs time slice observations, `gages_size` is non-zero, but gage_maxtimestep should be 0, so we would never enter that logic with the empty array of usgs_values. On the other hand, if there are values in usgs_values for most gages, but one or two of them are NaN, this check prevents us from filling the initial conditions array with useless NaN values (instead, we use the modeled value or the regular warm state.)